pkg/syncer/synctarget/gvr_source.go

```diff
@@ -0,0 +1,318 @@
+/*
+Copyright 2022 The KCP Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package synctarget
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	authorizationv1 "k8s.io/api/authorization/v1"
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsv1 "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/discovery/cached/memory"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/tools/cache"
+	"k8s.io/klog/v2"
+
+	kcpcorev1alpha1 "github.com/kcp-dev/kcp/pkg/apis/core/v1alpha1"
+	workloadv1alpha1 "github.com/kcp-dev/kcp/pkg/apis/workload/v1alpha1"
+	workloadv1alpha1informers "github.com/kcp-dev/kcp/pkg/client/informers/externalversions/workload/v1alpha1"
+	"github.com/kcp-dev/kcp/pkg/informer"
+)
+
+var _ informer.GVRSource = (*syncTargetGVRSource)(nil)
+
+// NewSyncTargetGVRSource returns an [informer.GVRSource] that can update its list of GVRs based on
+// a SyncTarget resource passed to the updateGVRs() method.
+//
+// It will be used to feed the various [informer.DiscoveringDynamicSharedInformerFactory] instances
+// for downstream and upstream.
+func NewSyncTargetGVRSource(
```

Comment:
this whole file surprises me in this PR. Why is there anything to do for GVRs when starting controllers?

Reply:
A change to the SyncTarget status (in either the `resourcesToSync` or the `virtualWorkspaces` or both) will impact either the GVRSource or the list of shards, or both.